README.md
nomnoml [![npm version](https://badge.fury.io/js/nomnoml.svg)](https://badge.fury.io/js/nomnoml) [![Known Vulnerabilities](https://snyk.io/test/npm/nomnoml/badge.svg)](https://snyk.io/test/npm/nomnoml)
=======

Hello, this is [nomnoml](http://www.nomnoml.com), a tool for drawing UML diagrams based on a simple syntax. It tries to keep its syntax visually as close as possible to the generated UML diagram without resorting to ASCII drawings.

Created by [Daniel Kallin](https://github.com/skanaar) and a cadre of [contributors](https://github.com/skanaar/nomnoml/graphs/contributors).

### Nomnoml was made possible by these cool projects

- [jison](http://zaach.github.io/jison/)
- [dagre](https://github.com/cpettitt/dagre)
- [typicons](http://typicons.com/)
- [codemirror](https://codemirror.net/)
- [solarized](http://ethanschoonover.com/solarized)

## Library

The [nomnoml](http://www.nomnoml.com) standalone javascript library can be used to render diagrams on your own web page. The only dependency is [dagre](https://github.com/dagrejs/dagre). Install it using either *npm* or good old script inclusion.

## SVG output in NodeJS

```
npm install nomnoml
```

```js
var nomnoml = require('nomnoml');
var src = '[nomnoml] is -> [awesome]';
console.log(nomnoml.renderSvg(src));
```

## HTML Canvas rendering target

```html
<script src="dagre.js"></script>
<script src="nomnoml.js"></script>

<canvas id="target-canvas"></canvas>
<script>
    var canvas = document.getElementById('target-canvas');
    var source = '[nomnoml] is -> [awesome]';
    nomnoml.draw(canvas, source);
</script>
```

## Command Line Interface

`npx nomnoml` exposes the SVG renderer with a command line interface. This mode also supports the `#import: <filename>` directive which allow you to separate complex diagrams into multiple files.

```
npx nomnoml input-file.noml
```

## Web application

The [nomnoml](http://www.nomnoml.com) web application is a simple editor with a live preview. It is purely client side and changes are saved to the browser's _localStorage_, so your diagram should be here the next time, (but no guarantees).

### Interaction

The canvas can be panned and zoomed by dragging and scrolling in the right hand third of the canvas. Downloaded image files will be given the filename in the `#title` directive.

### Example

This is how the Decorator pattern looks like in nomnoml syntax:

    [<frame>Decorator pattern|
      [<abstract>Component||+ operation()]
      [Client] depends --> [Component]
      [Decorator|- next: Component]
      [Decorator] decorates -- [ConcreteComponent]
      [Component] <:- [Decorator]
      [Component] <:- [ConcreteComponent]
    ]

### Association types

    -    association
    ->   association
    <->  association
    -->  dependency
    <--> dependency
    -:>  generalization
    <:-  generalization
    --:> implementation
    <:-- implementation
    +-   composition
    +->  composition
    o-   aggregation
    o->  aggregation
    --   note
    -/-  hidden
    _>   weightless edge
    __   weightless dashed edge

### Classifier types

    [name]
    [<abstract> name]
    [<instance> name]
    [<reference> name]
    [<note> name]
    [<package> name]
    [<frame> name]
    [<database> name]
    [<start> name]
    [<end> name]
    [<state> name]
    [<choice> name]
    [<input> name]
    [<sender> name]
    [<receiver> name]
    [<transceiver> name]
    [<actor> name]
    [<usecase> name]
    [<label> name]
    [<hidden> name]
    [<table> name| a | 5 || b | 7]

### Directives

    #import: my-common-styles.nomnoml
    #arrowSize: 1
    #bendSize: 0.3
    #direction: down | right
    #gutter: 5
    #edgeMargin: 0
    #gravity: 1
    #edges: hard | rounded
    #background: transparent
    #fill: #eee8d5; #fdf6e3
    #fillArrows: false
    #font: Calibri
    #fontSize: 12
    #leading: 1.25
    #lineWidth: 3
    #padding: 8
    #spacing: 40
    #stroke: #33322E
    #title: filename
    #zoom: 1
    #acyclicer: greedy
    #ranker: network-simplex | tight-tree | longest-path

### Custom classifier styles

A directive that starts with "." define a classifier style. The style is written as a space separated list of modifiers and key/value pairs.

    #.box: fill=#8f8 dashed
    #.blob: visual=ellipse title=bold
    [<box> GreenBox]
    [<blob> HideousBlob]

Modifiers

    dashed
    empty

Key/value pairs

    fill=(any css color)

    stroke=(any css color)

    align=center
    align=left

    direction=right
    direction=down

    visual=actor
    visual=class
    visual=database
    visual=ellipse
    visual=end
    visual=frame
    visual=hidden
    visual=input
    visual=none
    visual=note
    visual=package
    visual=receiver
    visual=rhomb
    visual=roundrect
    visual=sender
    visual=start
    visual=table
    visual=transceiver

Style title and text body

    title=left,italic,bold
    body=center,italic,bold
    
Text modifiers

    bold
    center
    italic
    left
    underline

## Contributing

If you want to contribute to the project more info is available in [CONTRIBUTING.md](CONTRIBUTING.md).


changelog.md
# Changelog

## [1.3.0]

- More options to style text
- Style node titles with `title=bold,italic,left`
- Style node bodys with `body=bold,italic,left`

## [1.2.0] - 2020-10-07

- Fixed layout bug when using HIDDEN nodes
- Fixed bug where SVG attributes were not properly escaped
- New association type `_>` that creates a shorter (weightless) edge
- New directive `#gravity: 0` that makes every association weightless
- Type definitions included (typescript .d file)
- Drop ES3 support and target ES5 explicitly

## [1.1.1] - 2020-09-21

- Fix bug when setting background color with `#background:`

## [1.1.0] - 2020-09-14

- Node name is attached to SVG shapes with a `data-name` attribute. Can be used for interactive diagrams.

## [1.0.1] - 2020-09-11

- Tweaked the text measurement heuristic for better handling of long texts.

## [1.0.0] - 2020-09-11

## Improved text measurement in SVG documents
- better heuristic for text width when no rendering context is available
- better text measurement when rendering context is available

## [0.10.0] - 2020-08-31

### New features
- Table is a new classifier type that displays its compartments in a uniform grid.
  Set the first row-break with a double pipe `||`.
- Non-boxy shapes are laid out properly. Ellipses, rhombuses, and cylinders (<usecase>, <choice>, and <database>) no longer intersect awkwardly with the arrows.

### Bug fixes
- fixed bug where SVG did not support bold text
- fixed bug where SVG did not track lineWidth in push/pop of graphic state

## [0.9.0] - 2020-08-26

- fixed bug where labels could extend outside bounding box

## [0.8.0] - 2020-07-09

- #background directive to specify background color

## [0.7.3] - 2020-07-08

- fixed incomplete rendering when edges extended outside bounding box

## [0.7.2] - 2020-05-14

- upgraded dependencies

## [0.7.1] - 2020-04-26

- fixed bug in compileFile()

## [0.7.0] - 2020-04-26

- expose compileFile to allow usage in a NodeJS environment

## [0.6.2] - 2019-10-27

- fix bug where SVG output did not get the correct text color

## [0.6.1] - 2019-07-22

- upgrade Lodash to fix a security vulnerability

## [0.6.0] - 2019-05-25

- fixed crash when two relations had the same start and end

## [0.5.0] - 2019-05-01

- improved text centering in SVG output

## [0.4.0] - 2019-04-26

### Library

- dagre library upgraded to 0.8.4
- version published as a `nomnoml.version` property

### www.nomnoml.com

- ie/edge support for downloading png images
- when new code results in parsing error, render last successful code.

### Known regressions

- self referential associations are broken


src/domain.ts
interface Config {
  padding: number
  stroke: string
  font: string
  title: string
  leading: number
  fontSize: number
  lineWidth: number
  gutter: number
  styles: { [key: string]: Style }
  fill: string[]
  background: string
  edges: string
  edgeMargin: number
  gravity: number
  spacing: number
  direction: 'TB'|'LR'
  fillArrows: boolean
  arrowSize: number
  bendSize: number
  zoom: number
  acyclicer: 'greedy' | undefined
  ranker: dagre.Ranker
}

interface Measurer {
    setFont(config: Config, isBold: 'bold'|'normal', isItalic:'italic'|'normal'): void
    textWidth(text: string): number
    textHeight(): number
}

interface Visualizer {
  (node: nomnoml.Classifier, x: number, y: number, config: Config, g: Graphics): void
}

interface NodeLayouter {
  (config: Config, node: nomnoml.Classifier): void
}

type Visual = 
  'actor'|
  'class'|
  'database'|
  'ellipse'|
  'end'|
  'frame'|
  'hidden'|
  'input'|
  'none'|
  'note'|
  'package'|
  'receiver'|
  'rhomb'|
  'roundrect'|
  'sender'|
  'start'|
  'table'|
  'transceiver'
  
interface TextStyle {
  bold: boolean
  underline: boolean
  italic: boolean
  center: boolean
}

interface Style {
  title: TextStyle
  body: TextStyle
  dashed: boolean
  empty: boolean
  fill: string|undefined
  stroke: string|undefined
  visual: Visual
  direction: 'TB'|'LR'|undefined
}
namespace nomnoml {

  export function buildStyle(conf: Partial<Style>, title: Partial<TextStyle>, body: Partial<TextStyle> = {}): Style {
    return {
      title: {
        bold: title.bold || false,
        underline: title.underline || false,
        italic: title.italic || false,
        center: title.center || false,
      },
      body: {
        bold: body.bold || false,
        underline: body.underline || false,
        italic: body.italic || false,
        center: body.center || false,
      },
      dashed: conf.dashed || false,
      empty: conf.empty || false,
      fill: conf.fill || undefined,
      stroke: conf.stroke || undefined,
      visual: conf.visual || 'class',
      direction: conf.direction || undefined,
    }
  }

  export class Compartment {
    x: number
    y: number
    width: number
    height: number
    offset: Vector
    constructor(
      public lines: string[], 
      public nodes: Classifier[], 
      public relations: Relation[]
    ){}
  }
  
  export interface RelationLabel {
    x?: number
    y?: number
    width?: number
    height?: number
    text: string
  }

  export class Relation {
    id: number
    path?: Vector[]
    start: string
    end: string
    startLabel: RelationLabel
    endLabel: RelationLabel
    assoc: string
  }

  export class Classifier implements Vec {
    x: number
    y: number
    width: number
    height: number
    layoutWidth: number
    layoutHeight: number
    dividers: Vector[][]
    constructor(
      public type: string,
      public name: string,
      public compartments: Compartment[]
    ){
      this.dividers = []
    }
  }
}


src/renderer.ts
namespace nomnoml {

	export function render(graphics: Graphics, config: Config, compartment: Compartment, setFont: nomnoml.SetFont){

		var g = graphics
		var vm = nomnoml.skanaar.vector

		function renderCompartment(compartment: Compartment, color: string, style: TextStyle, level: number){
			g.save()
			g.translate(compartment.offset.x, compartment.offset.y)
			g.fillStyle(color || config.stroke)
			compartment.lines.forEach(function (text, i){
				g.textAlign(style.center ? 'center' : 'left')
				var x = style.center ? compartment.width/2 - config.padding : 0
				var y = (0.5+(i+0.5)*config.leading)*config.fontSize
				if (text){
					g.fillText(text, x, y)
				}
				if (style.underline){
					var w = g.measureText(text).width
					y += Math.round(config.fontSize * 0.2)+0.5
					g.path([{x:x-w/2, y:y}, {x:x+w/2, y:y}]).stroke()
					g.lineWidth(config.lineWidth)
				}
			})
			g.translate(config.gutter, config.gutter)
			compartment.relations.forEach(function (r){ renderRelation(r) })
			compartment.nodes.forEach(function (n){ renderNode(n, level) })
			g.restore()
		}

		function renderNode(node: Classifier, level: number){
			var x = Math.round(node.x-node.width/2)
			var y = Math.round(node.y-node.height/2)
			var style = config.styles[node.type] || nomnoml.styles.CLASS

			g.fillStyle(style.fill || config.fill[level] || skanaar.last(config.fill))
			g.strokeStyle(style.stroke || config.stroke)
			if (style.dashed){
				var dash = Math.max(4, 2*config.lineWidth)
				g.setLineDash([dash, dash])
			}
			var drawNode = nomnoml.visualizers[style.visual] || nomnoml.visualizers.class
			g.setData('name', node.name)
			drawNode(node, x, y, config, g)
			g.setLineDash([])

			g.save()
			g.translate(x, y)

			node.compartments.forEach(function (part: Compartment, i: number){
				var textStyle = i == 0 ? style.title : style.body;
				if (style.empty) return
				g.save()
				g.translate(part.x, part.y)
				setFont(config, textStyle.bold ? 'bold' : 'normal', textStyle.italic ? 'italic' : undefined)
				renderCompartment(part, style.stroke, textStyle, level+1)
				g.restore()
			})
			for(var divider of node.dividers) {
				g.path(divider).stroke()
			}
			
			g.restore()
		}

		function strokePath(p: Vector[]){
			if (config.edges === 'rounded'){
				var radius = config.spacing * config.bendSize
				g.beginPath()
				g.moveTo(p[0].x, p[0].y)

				for (var i = 1; i < p.length-1; i++){
					g.arcTo(p[i].x, p[i].y, p[i+1].x, p[i+1].y, radius)
				}
				g.lineTo(skanaar.last(p).x, skanaar.last(p).y)
				g.stroke()
			}
			else
				g.path(p).stroke()
		}

		var empty = false, filled = true, diamond = true

	    function renderLabel(label: RelationLabel){
			if (!label || !label.text) return
			var fontSize = config.fontSize
			var lines = label.text.split('`')
			lines.forEach((l, i) => g.fillText(l, label.x, label.y + fontSize*(i+1)))
		}

		function renderRelation(r: Relation){
			var start = r.path[1]
			var end = r.path[r.path.length-2]
			var path = r.path.slice(1, -1)
			
			g.fillStyle(config.stroke)
			setFont(config, 'normal')

			renderLabel(r.startLabel)
			renderLabel(r.endLabel)

			if (r.assoc !== '-/-' && r.assoc !== '_/_'){
				if (skanaar.hasSubstring(r.assoc, '--') || skanaar.hasSubstring(r.assoc, '__')){
					var dash = Math.max(4, 2*config.lineWidth)
					g.setLineDash([dash, dash])
					strokePath(path)
					g.setLineDash([])
				}
				else
					strokePath(path)
			}

			function drawArrowEnd(id: string, path: Vector[], end: Vector){
				if (id === '>' || id === '<')
					drawArrow(path, filled, end, false)
				else if (id === ':>' || id === '<:')
					drawArrow(path, empty, end, false)
				else if (id === '+')
					drawArrow(path, filled, end, diamond)
				else if (id === 'o')
					drawArrow(path, empty, end, diamond)
			}

			var tokens = r.assoc.split(/[-_]/)
			drawArrowEnd(skanaar.last(tokens), path, end)
			drawArrowEnd(tokens[0], path.reverse(), start)
		}

		function drawArrow(path: Array<Vector>, isOpen: boolean, arrowPoint: Vector, diamond: boolean){
			var size = config.spacing * config.arrowSize / 30
			var v = vm.diff(path[path.length-2], skanaar.last(path))
			var nv = vm.normalize(v)
			function getArrowBase(s: number){ return vm.add(arrowPoint, vm.mult(nv, s*size)) }
			var arrowBase = getArrowBase(diamond ? 7 : 10)
			var t = vm.rot(nv)
			var arrowButt = (diamond) ? getArrowBase(14)
					: (isOpen && !config.fillArrows) ? getArrowBase(5) : arrowBase
			var arrow = [
				vm.add(arrowBase, vm.mult(t, 4*size)),
				arrowButt,
				vm.add(arrowBase, vm.mult(t, -4*size)),
				arrowPoint
			]
			g.fillStyle(isOpen ? config.stroke : config.fill[0])
			g.circuit(arrow).fillAndStroke()
		}

		function snapToPixels(){
			if (config.lineWidth % 2 === 1)
				g.translate(0.5, 0.5)
		}
		
		function setBackground() {
			g.clear()
			g.save()
			g.strokeStyle('transparent')
			g.fillStyle(config.background)
			g.rect(0, 0, compartment.width, compartment.height).fill()
			g.restore()
		}

		g.save()
		g.scale(config.zoom, config.zoom)
		setBackground()
		setFont(config, 'bold')
		g.lineWidth(config.lineWidth)
		g.lineJoin('round')
		g.lineCap('round')
		g.strokeStyle(config.stroke)
		snapToPixels()
		renderCompartment(compartment, undefined, buildStyle({}, {}).title, 0)
		g.restore()
	}
}


src/parser.ts
namespace nomnoml {
	interface ParsedDiagram {
		root: Compartment
		config: Config
	}
	export type AstRoot = AstCompartment
	type AstCompartment = AstSlot[]
	type AstSlot = string | AstClassifier | AstRelation
	type AstCompartmentList = AstCompartment[]
	interface AstRelation {
		assoc: string,
		start: AstClassifier,
		end: AstClassifier,
		startLabel: string,
		endLabel: string
	}
	interface AstClassifier { type: string, id: string, parts: AstCompartment[] }

	declare var nomnomlCoreParser: { parse(source: string): AstRoot }

	class Line {
			index: number
			text: string
	}

	export function parse(source: string): ParsedDiagram {
		function onlyCompilables(line: string){
			var ok = line[0] !== '#' && line.trim().substring(0,2) !== '//'
			return ok ? line.trim() : ''
		}
		function isDirective(line: Line): boolean { return line.text[0] === '#' }
		var lines: Line[] = source.split('\n').map(function (s, i){
			return {text: s, index: i }
		})
		var pureDirectives = lines.filter(isDirective)
		var directives: { [key: string]: string } = {}
		pureDirectives.forEach(function (line){
			try {
				var tokens =  line.text.substring(1).split(':')
				directives[tokens[0].trim()] = tokens[1].trim()
			}
			catch (e) {
				throw new Error('line ' + (line.index + 1) + ': Malformed directive')
			}
		})
		var pureDiagramCode = lines.map(e => onlyCompilables(e.text)).join('\n')
		
		if (pureDiagramCode == '') {
			return {
				root: new Compartment([], [], []),
				config: getConfig(directives)
			}
		}
		
		var parseTree = nomnoml.intermediateParse(pureDiagramCode)
		return {
			root: nomnoml.transformParseIntoSyntaxTree(parseTree),
			config: getConfig(directives)
		}

		function directionToDagre(word: any): 'TB'|'LR' {
			if (word == 'down') return 'TB'
			if (word == 'right') return 'LR'
			else return 'TB'
		}

		function parseRanker(word: string | undefined): dagre.Ranker {
			if (word == 'network-simplex' || word == 'tight-tree' || word == 'longest-path') {
				return word
			}
			return 'network-simplex'
		}

		function parseCustomStyle(styleDef: string): Style {
			var contains = skanaar.hasSubstring
			var floatingKeywords = styleDef.replace(/[a-z]*=[^ ]+/g, '')
			var titleDef = skanaar.last(styleDef.match('title=([^ ]*)') || [''])
			var bodyDef = skanaar.last(styleDef.match('body=([^ ]*)') || [''])
			return {
				title: {
					bold: contains(titleDef, 'bold') || contains(floatingKeywords, 'bold'),
					underline: contains(titleDef, 'underline') || contains(floatingKeywords, 'underline'),
					italic: contains(titleDef, 'italic') || contains(floatingKeywords, 'italic'),
					center: !(contains(titleDef, 'left') || contains(styleDef, 'align=left')),
				},
				body: {
					bold: contains(bodyDef, 'bold'),
					underline: contains(bodyDef, 'underline'),
					italic: contains(bodyDef, 'italic'),
					center: contains(bodyDef, 'center'),
				},
				dashed: contains(styleDef, 'dashed'),
				empty: contains(styleDef, 'empty'),
				fill: skanaar.last(styleDef.match('fill=([^ ]*)') || []),
				stroke: skanaar.last(styleDef.match('stroke=([^ ]*)') || []),
				visual: (skanaar.last(styleDef.match('visual=([^ ]*)') || []) || 'class') as Visual,
				direction: directionToDagre(skanaar.last(styleDef.match('direction=([^ ]*)') || [])),
			}
		}

		function getConfig(d: { [index:string]:string }): Config {
			var userStyles: { [index:string]: Style } = {}
			for (var key in d) {
				if (key[0] != '.') continue
				var styleDef = d[key]
				userStyles[key.substring(1).toUpperCase()] = parseCustomStyle(styleDef)
			}
			return {
				arrowSize: +d.arrowSize || 1,
				bendSize: +d.bendSize || 0.3,
				direction: directionToDagre(d.direction),
				gutter: +d.gutter || 5,
				edgeMargin: (+d.edgeMargin) || 0,
				gravity: +(d.gravity ?? 1),
				edges: d.edges == 'hard' ? 'hard' : 'rounded',
				fill: (d.fill || '#eee8d5;#fdf6e3;#eee8d5;#fdf6e3').split(';'),
				background: d.background || 'transparent',
				fillArrows: d.fillArrows === 'true',
				font: d.font || 'Helvetica',
				fontSize: (+d.fontSize) || 12,
				leading: (+d.leading) || 1.25,
				lineWidth: (+d.lineWidth) || 3,
				padding: (+d.padding) || 8,
				spacing: (+d.spacing) || 40,
				stroke: d.stroke || '#33322E',
				title: d.title || '',
				zoom: +d.zoom || 1,
				acyclicer: d.acyclicer === 'greedy' ? 'greedy' : undefined,
				ranker: parseRanker(d.ranker),
				styles: skanaar.merged(nomnoml.styles, userStyles)
			};
		}
	}

	export function intermediateParse(source: string): AstRoot {
		return nomnomlCoreParser.parse(source)
	}

	export function transformParseIntoSyntaxTree(entity: AstRoot): Compartment {

		function isAstClassifier(obj: AstSlot): obj is AstClassifier {
			return (<AstClassifier>obj).parts !== undefined
		}

		function isAstRelation(obj: AstSlot): obj is AstRelation {
			return (<AstRelation>obj).assoc !== undefined
		}

		function isAstCompartment(obj: any): obj is AstCompartment {
			return Array.isArray(obj)
		}

		var relationId: number = 0

		function transformCompartment(slots: AstCompartment): Compartment {
			var lines: string[] = []
			var rawClassifiers: AstClassifier[] = []
			var relations: Relation[] = []
			slots.forEach(function (p: AstSlot){
				if (typeof p === 'string')
					lines.push(p)
				if (isAstRelation(p)){ // is a relation
					rawClassifiers.push(p.start)
					rawClassifiers.push(p.end)
					relations.push({
							id: relationId++,
							assoc: p.assoc,
							start: p.start.parts[0][0] as string,
							end: p.end.parts[0][0] as string,
							startLabel: { text: p.startLabel },
							endLabel: { text: p.endLabel }
						})
					}
				if (isAstClassifier(p)){
					rawClassifiers.push(p)
				}
			})
			var allClassifiers: Classifier[] = rawClassifiers
				.map(transformClassifier)
				.sort(function(a: Classifier, b: Classifier): number {
					return b.compartments.length - a.compartments.length
				})
			var uniqClassifiers = skanaar.uniqueBy(allClassifiers, 'name')
			var uniqRelations = relations.filter(function (a){
				for (var b of relations) {
					if (a === b) return true
					if (b.start == a.start && b.end == a.end) return false
				}
				return true
			})
			return new Compartment(lines, uniqClassifiers, uniqRelations)
		}

		function transformClassifier(entity: AstClassifier): Classifier {
				var compartments = entity.parts.map(transformCompartment)
				return new Classifier(entity.type, entity.id, compartments)
		}

		return transformCompartment(entity)
	}
}
